### PR TITLE
[21.11] nixStable: 2.4 -> 2.3.16

### PR DESCRIPTION
**Pull request by tomberek**
Revert to 2.3.16

###### Motivation for this change
https://discourse.nixos.org/t/nix-2-4-and-what-s-next/16257

###### Things done
Revert and add release notes discussing the change.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by tomberek** [comment]
Alias nixFlakes to nixUnstable to ensure the named version actually supports flakes.

Additional relevant issues:
#145237
#147485 
#147387

File: pkgs/top-level/aliases.nix [hunk 1]
```diff
@@ -600,7 +600,7 @@ mapAliases ({
   nilfs_utils = nilfs-utils; # added 2018-04-25
   nix-direnv-flakes = nix-direnv;
   nix-review = nixpkgs-review; # added 2019-12-22
-  nixFlakes = nixStable; # added 2021-05-21
+  nixFlakes = nixUnstable; # added 2021-05-21
```
Comment (thiagokokada) [inline pkgs/top-level/aliases.nix:603]: Why not `nix_2_4` :thinking: ?

**Review by thiagokokada — COMMENTED** [review]
Well, I am against this change, mostly because those issues seems to be happening with leaf, non-critical packages.

Also, Nix 2.3 is available as `nix_2_3` for those who need.

File: pkgs/top-level/aliases.nix (hunk 1 above)
Comment (tomberek) [reply pkgs/top-level/aliases.nix]: That does make a bit more sense. Thx.

File: nixos/doc/manual/release-notes/rl-2111.section.md [hunk 2]
```diff
@@ -6,7 +6,7 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 ## Highlights {#sec-release-21.11-highlights}
 
-- Nix has been updated to version 2.4, reference its [release notes](https://discourse.nixos.org/t/nix-2-4-released/15822) for more information on what has changed. The previous version of Nix, 2.3.16, remains available for the time being in the `nix_2_3` package.
+- Nix version remains at 2.3.16. Nix has not been updated to version 2.4 due to regressions in non-experimental behavior. To upgrade to 2.4, use `nixos-unstable` branch or `nixUnstable`/`nix_2_4` attributes; reference the [release notes](https://discourse.nixos.org/t/nix-2-4-released/15822) for more information on upcoming changes. Provide breakage reports at [https://github.com/NixOS/nix/issues]()
```
Comment (samueldr) [inline nixos/doc/manual/release-notes/rl-2111.section.md:9]:
```suggestion
- Nix version remains at 2.3.16. Nix has not been updated to version 2.4 due to regressions in non-experimental behavior. To upgrade to 2.4, use the `nixos-unstable` branch or set the `nix.package` options to either the `nixUnstable` or `nix_2_4` attributes. Read the [release notes](https://discourse.nixos.org/t/nix-2-4-released/15822) for more information on upcoming changes. Please help us improving Nix by providing any breakage reports at [https://github.com/NixOS/nix/issues]()
```

Don't take the whole rewrite as the only way to change this. Main thing to consider is making better structured sentences, *and importantly* to tell which option to set to use Nix 2.4.

**Review by samueldr — DISMISSED** [review]

**Comment by tomberek** [comment]
@ofborg eval

**Comment by bobby285271** [comment]
The https://github.com/NixOS/nix/issues link in release note doesn't work.

If the revert has been decided, I guess at least things like 90c1cdd93f4f1e68a70a9bc8ad0185fc97be5923 and fd4390146e2ccb043bbb0a07c1cd89fe01ee1315 also needs to be reverted to build nixpkgs-lib-tests, and also probably other changes made in #144197 and other PRs that I don't know?

**Comment by thiagokokada** [comment]
I think it would be good to involve the folks involved with the bump here. Not everyone (me included) participates in Discourse discussions often, and I think the topic in Discourse has a low number of inputs from other folks.

CC @lovesegfault @roberth @siraben @SuperSandro2000 @reckenrode

**Review by nrdxp — APPROVED** [review]
I don't see any good option besides moving forward with this. I can wait a bit before merging to let some others give their input, but having a Nix which requires experimental features in a stable OS release in order to function without buggy, surprising, or annoying behavior doesn't seem tenable.

I think, by the next release, we should have a few patch releases to help fix up the 2.4 line, and hopefully we can bump it by then. For now, I feel this is the best move.

**Comment by reckenrode** [comment]
Would the version of Nix available for download from nixos.org be reverted as well? The 2.4 installer has a number of fixes and improvements for Darwin. I’m pretty sure the 2.3 installer doesn’t even work on macOS 12 due to changes Apple made that broke the installer’s read-only root detection logic.

**Comment by nrdxp** [comment]
@reckenrode, this change is only meant to affect 21.11, so, somebody correct me if I'm wrong, but I don't think we will be reverting the installer available for Nix use on other systems.

**Comment by thiagokokada** [comment]
> @reckenrode, this change is only meant to affect 21.11, so, somebody correct me if I'm wrong, but I don't think we will be reverting the installer available for Nix use on other systems.

I think that having the newest release of NixOS using an older version of Nix, while the official site have the latest version can also be surprising. Maybe we should at least have a note somewhere explaining why, for example, a pinned issue on GitHub?

**Comment by Ma27** [comment]
> Maybe we should at least have a note somewhere explaining why, for example, a pinned issue on GitHub?

I think it's OK to document this in the release-notes, that's what they are for :)

**Comment by vcunat** [comment]
This broke `nixosTests.misc`, e.g. https://hydra.nixos.org/build/159776540

**Comment by vcunat** [comment]
I assume we also need to revert commit 546d60c5e ?

**Comment by bobby285271** [comment]
Shall we also revert #145437 (for `nixos/test/boot`)?

**Comment by bobby285271** [comment]
> Shall we also revert #145437 (for `nixos/test/boot`)?

I opened #147789 for this.

**Comment by vcunat** [comment]
It seems that `nixosTests.boot.*` do time out consistently, so something like that is probably needed.

**Comment by vcunat** [comment]
For reference, the `tested` job succeeded after PR #147789  🎉

**Comment by Mic92** [comment]
Please also review https://github.com/NixOS/nixpkgs/pull/147998 which was not addressed in this PR.

**Comment by nbp** [comment]
Shouldn't this change be also applied to nixos-unstable channel?

**Comment by roberth** [comment]
@nbp I hope that we'll have a minor release that improves the situation. I don't think reverting nixos-unstable to nix 2.3 is part of a sustainable solution.

**Pull request merged**